Unique ids generator package

evridiki91 opened the pull request:
Unique ids generator

dulciejackson commented on packages/backend/unique-ids-generator/README.md:
```diff
@@ -0,0 +1,21 @@
+  <img src="https://img.shields.io/npm/v/@gravitywelluk/unique-ids-generator" alt="Version" />
+</p>
+<br />
+
```
Can we add a description of what this generates by default?
For instance:
* Examples of the kind of ID that's generated
* A ballpark of _how_ random these are - e.g. how many different 4 character IDs are there?

dulciejackson commented on packages/backend/unique-ids-generator/__tests__/unique-ids-generator.test.ts:
```diff
@@ -0,0 +1,30 @@
+    const LETTERS_ONLY = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+    const uniqueId = generateUniqueId({ dictionary: LETTERS_ONLY });
+
+    expect(/^[A-Za-z]*$/.test(uniqueId)).toBeTruthy();
```
Can we also check that it's falsy for numbers?

evridiki91 replied:
If it contains numbers `.test()` is going to return false!

dulciejackson commented on packages/backend/unique-ids-generator/src/generate-unique-id.ts:
```diff
@@ -0,0 +1,32 @@
+
+const debug = createDebug("UNIQUE_IDS_GENERATOR:GENERATE-UNIQUE-ID");
+
+interface GenerateUniqIdParams {
```
Should this be `GenerateUniqueIdParams`?

evridiki91 replied:
Good point !

dulciejackson reviewed (commented):
Looking good! 🦞 

A couple of things: 
* What happens if you provide the same character more than once in the dictionary? I'd imagine that we'd ideally want to dedupe the character set before using it, to be sure that:
  1. It doesn't break the function (which I don't think it will)
  2. It doesn't undermine the random-ness, by skewing the likelihood of a specific character if it's provided > 1 time in the string

* Note that `Math.random` is **pseudo**-random, so whilst its values appear to be random, they're not quite. It roughly relies on timestamps to generate its values, which means that there's a non-zero chance of generating the same ID at the same time for two users - given that more than one could be registering at once. See the third answer here: https://stackoverflow.com/questions/2344312/how-is-randomness-achieved-with-math-random-in-javascript.

Of course, both of these are probably fine for our purposes for now, but we may want a disclaimer in the README here (and maybe in a JSDoc comment above the function) so we're aware of the caveats when using this on other projects!

evridiki91 commented:
@dulciejackson I've update the generator to always remove duplicates from the dictionary (also added an extra test case for that scenario) and updated the README file to include information about everything you recommended above 🦞
